src/sandstorm/supervisor.c++

```diff
@@ -1824,11 +1859,20 @@ public:
     // # Wraps `MainView.restore()`. Can also restore capabilities hosted by the supervisor.
     auto objectId = context.getParams().getRef();
 
-    if (objectId.which() == SupervisorObjectId<>::WAKE_LOCK_NOTIFICATION) {
-      context.getResults().setCap(wakelockSet.restore(objectId.getWakeLockNotification()));
-      return kj::READY_NOW;
-    } else {
-      KJ_FAIL_REQUIRE("Supervisor can only restore wakelocks for now.");
+    switch (objectId.which()) {
+      case SupervisorObjectId<>::WAKE_LOCK_NOTIFICATION: {
+        context.getResults().setCap(wakelockSet.restore(objectId.getWakeLockNotification()));
+        return kj::READY_NOW;
+      }
+      case SupervisorObjectId<>::APP_REF: {
+        auto req = mainView.castAs<MainView<>>().restoreRequest();
```

Comment:
I think we actually want `mainView` to be of type `MainView<>::Client` all along rather than `castAs` here.

This brings up another issue, which is that technically the supervisor is supposed to filter requests to this capability such that `MainView`'s methods are not visible to outsiders (but `UiView`'s methods are).